packages/devtools/lib/src/timeline/timeline_protocol.dart

```diff
@@ -166,120 +169,201 @@ class TimelineData {
+  ///
+  /// Returns a bool indicating whether the event was added to the frame.
+  bool _maybeAddEventToFrame(TimelineEvent e, TimelineFrame f) {
+    assert(f.wellFormed);
 
-    final TimelineEvent current = _asyncEvents[asyncUID];
-    if (current != null) {
-      current.endTime = event.timestampMicros;
-      _asyncEvents[asyncUID] = current.parent;
+    // TODO(kenzie): consider trimming VSYNC layer from pipelineProduceFlow. It
+    // can start outside of the frame's time boundaries and could pose a risk
+    // for us missing a frame.
 
-      if (_asyncEvents[asyncUID] == null) {
-        frames[event.frameId].addEvent(current);
-      }
+    // Ensure the event fits within the frame's time boundaries.
+    if (!_eventOccursWithinFrameBoundaries(e, f)) return false;
+
+    bool eventAdded = false;
+
+    if (e.isPipelineProduceFlow && f.pipelineProduceFlow == null) {
+      f.pipelineProduceFlow = e;
+      eventAdded = true;
+    } else if (e.isPipelineConsumeFlow && f.pipelineConsumeFlow == null) {
+      f.pipelineConsumeFlow = e;
+      eventAdded = true;
     }
+
+    // Adding event 'e' could mean we have completed the frame. Check if we
+    // should add the completed frame to [_frameCompleteController].
+    _maybeAddCompletedFrame(f);
+
+    return eventAdded;
   }
 
-  TimelineEventType _getTimelineEventType(TraceEvent event) {
-    // The event will only be of type 'cpu' or 'gpu' because we do not process
-    // events of any other type.
-    if (event.isCpuEvent) {
-      return TimelineEventType.cpu;
-    } else {
-      return TimelineEventType.gpu;
+  bool _eventOccursWithinFrameBoundaries(TimelineEvent e, TimelineFrame f) {
+    // Epsilon in microseconds.
+    const int epsilon = 50;
+
+    // Allow the event to extend the frame boundaries by `epsilon` microseconds.
+    final bool fitsStartBoundary = f.startTime - e.startTime - epsilon < 0;
+    final bool fitsEndBoundary = f.endTime - e.endTime + epsilon > 0;
+    return fitsStartBoundary && fitsEndBoundary;
+  }
+
+  List<TimelineFrame> _getAndSortWellFormedFrames() {
+    final List<TimelineFrame> frames =
+        List<TimelineFrame>.from(_pendingFrames.values)
+            .where((TimelineFrame f) => f.wellFormed)
+            .toList();
+
+    // Sort frames by their startTime. Sorting these frames ensures we will
+    // handle the oldest frame first when iterating through the list.
+    frames.sort((a, b) => a.startTime.compareTo(b.startTime));
+    return frames;
+  }
+
+  void _maybeAddCompletedFrame(TimelineFrame frame) {
+    if (frame.readyForTimeline && !frame.addedToTimeline.isCompleted) {
+      _frameCompleteController.add(frame);
+      _pendingFrames.remove(frame);
+      frame.addedToTimeline.complete();
     }
   }
 }
 
+/// Data describing a single frame.
+///
+/// Each TimelineFrame should have 4 distinct pieces of data:
+///   - [_pipelineProduceFlow] : flow of events showing the CPU work for the
```

Comment:
reference public members instead of private where possible in dartdoc comments.

Reply:
Generally I think * should be used instead of - in dartdoc to indicate bullet items.

Reply:
done.